packages/angular_devkit/build_angular/src/browser/index.ts

```diff
@@ -359,30 +370,117 @@ export function buildWebpackBrowser(
 
               // Execute the bundle processing actions
               context.logger.info('Generating ES5 bundles for differential loading...');
-              await new Promise<void>((resolve, reject) => {
-                const workerFile = require.resolve('../utils/process-bundle');
-                const workers = workerFarm(
-                  {
-                    maxRetries: 1,
-                  },
-                  path.extname(workerFile) !== '.ts'
-                    ? workerFile
-                    : require.resolve('../utils/process-bundle-bootstrap'),
-                  ['process'],
-                );
-                let completed = 0;
-                const workCallback = (error: Error | null) => {
-                  if (error) {
-                    workerFarm.end(workers);
-                    reject(error);
-                  } else if (++completed === actions.length) {
-                    workerFarm.end(workers);
-                    resolve();
+
+              const processActions: typeof actions = [];
+              const cacheActions: { src: string; dest: string }[] = [];
+              for (const action of actions) {
```

Comment:
I am not fully grasping the need for parts of these logic statements. Maybe I am missing some use case or something important that I am not thinking of.

But given an input the output should always be the same optimization and sourcemap settings. Because the input files contents will differ based on the following build settings:
- sourceMaps: true, false, hidden or vendor
- script optimization: true or false
- optimizeOnly: true or false

Each of the above setting will create different content in the original file (.map or .js) and hence when generating the file hash it should be enough.

Hence, I would picture the logic to take into consideration both file contents which should also aid a bit readability.

Something along;
```ts
const cacheActions: { src: string; dest: string }[] = [];

for (const action of actions) {
  const codeHash = createHash('sha1')
    .update(action.code)
    .digest('hex');
  const cacheKey = `${packageVersion}-${action.code.length}-${codeHash}`;

  let mapCacheKey: undefined | string;
  if (action.map && action.sourceMaps) {
    const mapHash = createHash('sha1')
      .update(action.map)
      .digest('hex');

    mapCacheKey = `${packageVersion}-${action.map.length}-${mapHash}`;
  }

  const originalEntry = await cacache.get.info(cacheDownlevelPath, `${cacheKey}-orig`);
  const originalMapEntry = mapCacheKey && (await cacache.get.info(cacheDownlevelPath, `${mapCacheKey}-orig-map`));
  const isOriginalCached = originalEntry && (originalMapEntry || !action.sourceMaps);

  // if original file is not cached it's downleveled or the optimized bundle is not cached either
  if (isOriginalCached) {
    if (actions.optimizeOnly) {
      // bundle only needs to be optimized
      cacheActions.push({ src: originalEntry.path, dest: action.file });
      if (originalMapEntry) {
        cacheActions.push({ src: originalMapEntry.path, dest: `${action.file}.map` });
      }

      continue;
    } else {
      // Bundle needs to be downleveled
      const downlevelEntry = await cacache.get.info(cacheDownlevelPath, `${cacheKey}-dl`);
      const downlevelMapEntry = mapCacheKey && (await cacache.get.info(cacheDownlevelPath, `${mapCacheKey}-dl-map`));
      const isDownLeveledCached = downlevelEntry && (downlevelMapEntry || !action.sourceMaps);

      if (isDownLeveledCached) {
        const newFileName = action.filename.replace('-es2015.', '-es5.');
        cacheActions.push({ src: downlevelEntry.path, dest: newFileName });
        if (downlevelMapEntry) {
          cacheActions.push({ src: downlevelMapEntry.path, dest: `${newFileName}.map` });
        }

        continue;
      }
    }
  }

  processActions.push({
    ...action,
    cacheKey,
    mapCacheKey,
    cachePath: cacheDownlevelPath || undefined,
  });
}

for (const action of cacheActions) {
  fs.copyFileSync(action.src, action.dest, fs.constants.COPYFILE_FICLONE);
}
```

Reply:
Completely reworked the logic to greatly simplify it as well as minimizing the number of cache access attempts to just the required entries.
I was avoiding hashing the source map files as they tend to be quite large.

Reply:
Much better :)